### PR TITLE
improve drain check

### PR DESCRIPTION
**Pull request by SchSeba**
We need to check in the drain function also if VFs exist
If they exist we need to check if any other configuration is required (link driver and MTU)

If a change to the VF is requested we should drain the node because we can have pods using the VF

Signed-off-by: Sebastian Sch <sebassch@gmail.com>

### CODE REVIEW

**Comment by zeeke** [comment]
I think there is a regression in case the `desired` interface Mtu is 0.

I arranged a unit test with information taken from a test cluster here:
https://github.com/zeeke/sriov-network-operator-1/blob/pr354-test/pkg/plugins/generic/generic_plugin_test.go

It works without changes at [generic_plugin.go:166](https://github.com/k8snetworkplumbingwg/sriov-network-operator/pull/354/files#diff-5e18b1bb6723228bb51dcd9afe17e63609c26b293a04e6f0a44075ef0f5cfcc2L166).

Can you please add the test to the PR?

**Comment by zeeke** [comment]
LGTM

File: pkg/plugins/generic/generic_plugin.go [hunk 1]
```diff
@@ -163,8 +163,8 @@ func needDrainNode(desired sriovnetworkv1.Interfaces, current sriovnetworkv1.Int
 					needDrain = true
 					return
 				}
-				if iface.Mtu != 0 && iface.Mtu != ifaceStatus.Mtu {
-					glog.V(2).Infof("generic-plugin needDrainNode(): need drain, expect MTU %v, current MTU %v", iface.Mtu, ifaceStatus.Mtu)
+				if utils.NeedUpdate(&iface, &ifaceStatus) {
```
Comment (adrianchiris) [inline pkg/plugins/generic/generic_plugin.go:166]: this method checks also num VFs so you can remove the condition at L161

File: pkg/plugins/generic/generic_plugin_test.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
+package generic
+
+import (
+	"testing"
+
+	sriovnetworkv1 "github.com/k8snetworkplumbingwg/sriov-network-operator/api/v1"
+	"github.com/stretchr/testify/assert"
+)
+
```
Comment (adrianchiris) [inline pkg/plugins/generic/generic_plugin_test.go:9]: Any chance to use ginkgo and gomega as we do in other places  ?

also you could define a test package (i.e `generic_test`) and just test the interface calls WDYT ?

File: pkg/plugins/generic/generic_plugin.go (hunk 1 above)
Comment (SchSeba) [reply pkg/plugins/generic/generic_plugin.go:161]: done

File: pkg/plugins/generic/generic_plugin_test.go (hunk 2 above)
Comment (SchSeba) [reply pkg/plugins/generic/generic_plugin_test.go:11]: done

**Comment by SchSeba** [comment]
Hi @adrianchiris please take a look when you have time

**Review by e0ne — APPROVED** [review]
/LGTM

File: pkg/plugins/generic/generic_plugin_test.go [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+			needDrain, needReboot, err := genericPlugin.OnNodeStateChange(networkNodeState)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(needReboot).To(BeFalse())
+			Expect(needDrain).To(BeTrue())
+		})
+	})
+
+})
+
+//func TestNeedDrainNode_InterfaceWithBadMtuShouldDrainTheNode(t *testing.T) {
```
Comment (adrianchiris) [inline pkg/plugins/generic/generic_plugin_test.go:133]: nit : delete ?

**Review by adrianchiris — APPROVED** [review]
Minor nit. and im LGTM

File: pkg/plugins/generic/generic_plugin_test.go (hunk 3 above)
Comment (SchSeba) [reply pkg/plugins/generic/generic_plugin_test.go]: miss that :P

Thanks!

**Review by adrianchiris — APPROVED** [review]

**Pull request merged**